[Navigation] Implement more of NavigationHistoryEntry

rwlbuis opened the pull request:
#### 07ca7d40e32bb25bfc5ba2bc2e41870d0b20e5bc
<pre>
[Navigation] Implement more of NavigationHistoryEntry
<a href="https://bugs.webkit.org/show_bug.cgi?id=268986">https://bugs.webkit.org/show_bug.cgi?id=268986</a>

Reviewed by Alex Christensen.

Implement more of NavigationHistoryEntry according to the specification:
- implement index by looking the NavigationHistoryEntry up in the entries of the associated Navigation object.
- implement sameDocument by checking whether it is same document to the current history item.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/current-basic-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::index const):
(WebCore::NavigationHistoryEntry::sameDocument const):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: <a href="https://commits.webkit.org/275153@main">https://commits.webkit.org/275153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfc354cac7f670ab8368cc5e93f1ed07d8a7a4b2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19912 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43457 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36990 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22918 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17243 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33897 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41473 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35254 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14520 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14626 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36244 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44752 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37111 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40305 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38674 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17333 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9206 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17384 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16977 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -104,7 +104,7 @@ EventTargetInterface Navigation::eventTargetInterface() const
 {
     // FIXME: keep track of promises to resolve later.
-    Ref entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
+    Ref entry = NavigationHistoryEntry::create(scriptExecutionContext(), { }, -1);
```
-1 seems a bit magical. Would this be better as std::optional or some such?

rwlbuis replied:
Good point, changed to use std::optional indeed.

achristensen07 commented on Source/WebCore/page/NavigationHistoryEntry.cpp:
```diff
@@ -63,6 +68,30 @@ EventTargetInterface NavigationHistoryEntry::eventTargetInterface() const
+{
+    RefPtr document = dynamicDowncast<Document>(scriptExecutionContext());
+    if (!document || !document->domWindow())
+        return -1;
```
Return a std::optional<uint64_t> and return std::nullopt here?

rwlbuis replied:
Unfortunately this is dictated by the spec.

achristensen07 commented on Source/WebCore/page/NavigationHistoryEntry.h:
```diff
@@ -71,10 +70,9 @@ class NavigationHistoryEntry final : public RefCounted<NavigationHistoryEntry>,
     // TODO: Every entry is supposed to have an associated history item.
     std::optional<Ref<HistoryItem>> m_associatedHistoryItem;
-    bool m_sameDocument;
+    std::optional<long long> m_documentSequenceNumber;
```
Does this need to be signed?  I think it should be std::optional<uint64_t>

rwlbuis replied:
I would like to match what HistoryItem is using for the documentSequenceNumber, which uses signed.

rwlbuis committed:
[Navigation] Implement more of NavigationHistoryEntry
https://bugs.webkit.org/show_bug.cgi?id=268986

Reviewed by Alex Christensen.

Implement more of NavigationHistoryEntry according to the specification:
- implement index by looking the NavigationHistoryEntry up in the entries of the associated Navigation object.
- implement sameDocument by checking whether it is same document to the current history item.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/current-basic-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::index const):
(WebCore::NavigationHistoryEntry::sameDocument const):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: https://commits.webkit.org/275153@main
